synapse/tools/storm.py

```diff
@@ -215,6 +219,202 @@ async def runCmdOpts(self, opts):
         except s_exc.SynErr as e:
             self.printf(e.errinfo.get('mesg', str(e)))
 
+def cmplgenr(*genrs, prefix=''):
+    '''
+    Iterate over all the generators/iterators passed in as args and return Completions from them.
+    If prefix is specified, make sure the current item in the generator starts with the prefix value.
+    '''
+    for genr in genrs:
+        for (valu, display) in genr:
+            if prefix and not valu.startswith(prefix):
+                continue
+
+            completion = prompt_toolkit.completion.Completion(
+                valu[len(prefix):],
+                display=display
+            )
+
+            yield completion
+
+tagre = regex.compile(r'#(\w+[\w\.]*)$')
+libre = regex.compile(r'\$([a-z_][a-zA-Z0-9_\.]*)$')
+cmdpropre = regex.compile(r'([a-z_][a-z0-9_]+[a-z0-9_:\.]+)$')
+
+class StormCompleter(prompt_toolkit.completion.Completer):
+    def __init__(self, cli):
+        self._cli = cli
+
+        self.initialized = False
+
+        # These are all the possible completions. Their format should be as follows:
+        # (<name>, <display>)
+
+        # ('misp.event.add', '[cmd] misp.event.add - Add Synapse nodes to a MISP server.')
+        self._cmds = []
+
+        # ('lib.cast', '[lib] $lib.cast(name: str, valu: any) - Normalize a value as a Synapse Data Model Type.')
+        # ('lib.debug', '[lib] $lib.debug - True if the current runtime has debugging enabled.')
+        self._libs = []
+
+        # ('inet:fqdn', '[form] inet:fqdn - A Fully Qualified Domain Name (FQDN).')
+        # ('inet:fqdn:domain', '[prop] inet:fqdn:domain - The parent domain for the FQDN.')
+        self._forms = []
+        self._props = []
+
+    async def load(self):
+        # Process forms/props
+        q = '''
+        $forms = ([])
+        $props = ([])
+        syn:prop
+        { -:relname $forms.append(($node.repr(), :doc)) }
+        { +:relname $props.append(($node.repr(), :doc)) }
+        fini {
+            return(($forms, $props))
+        }
+        '''
+        forms, props = await self._cli.item.callStorm(q)
```

Comment:
We should be able to get forms / props via the `getModelDict()` API?

Reply:
sure.

Reply:
sure